pilot/pkg/networking/core/v1alpha3/envoyfilter/rc_patch.go

```diff
@@ -95,7 +80,82 @@ func ApplyRouteConfigurationPatches(patchContext networking.EnvoyFilter_PatchCon
 		}
 		routeConfiguration.VirtualHosts = trimmedVirtualHosts
 	}
-	return routeConfiguration
+}
+
+func doVirtualHostOperation(proxy *model.Proxy, patchContext networking.EnvoyFilter_PatchContext,
+	patches map[networking.EnvoyFilter_ApplyTo][]*model.EnvoyFilterConfigPatchWrapper,
+	routeConfiguration *xdsapi.RouteConfiguration, virtualHost *route.VirtualHost, virtualHostRemoved *bool) {
+
+	for _, cp := range patches[networking.EnvoyFilter_VIRTUAL_HOST] {
+		if commonConditionMatch(proxy, patchContext, cp) &&
+			routeConfigurationMatch(patchContext, routeConfiguration, cp) &&
+			virtualHostMatch(virtualHost, cp) {
+
+			if cp.Operation == networking.EnvoyFilter_Patch_REMOVE {
+				virtualHost.Name = ""
+				*virtualHostRemoved = true
+				// nothing more to do.
+				return
+			} else if cp.Operation == networking.EnvoyFilter_Patch_MERGE {
+				proto.Merge(virtualHost, cp.Value)
+			}
+		}
+	}
+	doHTTPRouteListOperation(proxy, patchContext, patches, routeConfiguration, virtualHost)
+}
+
+func doHTTPRouteListOperation(proxy *model.Proxy, patchContext networking.EnvoyFilter_PatchContext,
+	patches map[networking.EnvoyFilter_ApplyTo][]*model.EnvoyFilterConfigPatchWrapper,
+	routeConfiguration *xdsapi.RouteConfiguration, virtualHost *route.VirtualHost) {
+
+	routesRemoved := false
+	// Apply the route level removes/merges if any.
+	for index := range virtualHost.Routes {
+		doHTTPRouteOperation(proxy, patchContext, patches, routeConfiguration, virtualHost, index, &routesRemoved)
+	}
+
+	// now for the adds
+	for _, cp := range patches[networking.EnvoyFilter_HTTP_ROUTE] {
+		if cp.Operation != networking.EnvoyFilter_Patch_ADD {
+			continue
+		}
+		if commonConditionMatch(proxy, patchContext, cp) &&
+			routeConfigurationMatch(patchContext, routeConfiguration, cp) &&
```

Comment:
no need `routeConfigurationMatch` as `virtualHostMatch` can do that?

Reply:
ignore, they checks different

Reply:
no. They are all independent of each other.. the virtualHostMatch does not check inside the routeConfigurationMatch fields